source/ShipsFactory.cpp

```diff
@@ -0,0 +1,385 @@
+/* ShipsFactory.cpp
+Copyright (c) 2022 by Peter van der Meer
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "ShipsFactory.h"
+
+#include "DataNode.h"
+#include "Files.h"
+#include "GameData.h"
+#include "SpriteSet.h"
+#include "UniverseObjects.h"
+
+#include <algorithm>
+
+using namespace std;
+
+namespace {
+	// Get an overview of how many weapon-outfits are equipped.
+	map<const Outfit *, int> GetEquipped(const vector<Hardpoint> &weapons)
+	{
+		map<const Outfit *, int> equipped;
+		for(const Hardpoint &hardpoint : weapons)
+			if(hardpoint.GetOutfit())
+				++equipped[hardpoint.GetOutfit()];
+		return equipped;
+	}
+}
+
+
+
+ShipsFactory::ShipsFactory(UniverseObjects &universe): universe(universe)
+{
+}
+
+
+
+// Load a ship from a datafile. Creation of the Ship object itself has
+// already been done by the caller. Getting only the DataNode as parameter
+// and returning a shared_ptr<Ship> is nicer than getting the ship as
+// parameter by reference, but the ES code allows overwriting ship-
+// definitions by a new load, so we need to support overwriting existing
+// ship definitions here.
+void ShipsFactory::LoadShip(Ship &ship, const DataNode &node) const
+{
+	if(node.Size() >= 2)
+	{
+		ship.SetModelName(node.Token(1));
+		ship.pluralModelName = node.Token(1) + 's';
+	}
+	if(node.Size() >= 3)
+	{
+		ship.base = GameData::Ships().Get(ship.ModelName());
+		ship.variantName = node.Token(2);
+	}
+	ship.isDefined = true;
```

Comment:
Right, we might even want to delay the creation of the "runtime" data models until all other loading is done, so that we don't even need a FinishLoading() anymore.